keps/sig-node/20190322-image-gc-whitelist.md

```diff
@@ -0,0 +1,273 @@
+Whitelisted images will be ignored during garbage collection.
+Currently only the dockershim pod sandbox image is whitelisted, this KEP aims
+to make the whitelist configurable.
+
+## Motivation
+
+There are some use cases when not all the container images are available from a remote repository, therefore container images are stored locally in the image cache of the Nodes. These container images can be preloaded to the Nodes instead of pulled on demand.
+The image garbage collector in Kubelet deletes all unused images from the image cache in case of storage shortage.
+When a Pod using such preloaded image scheduled to a Node from where the container image was garbage collected the Pod will go to ImagePullBackOff state forever.
+
+Users of container runtimes other than dockershim may wish to whitelist the
+podsandbox image used by their container runtime, which may not be in sync
+with dockershim.
+
+### Goals
+
+- Add a Kubelet command line option to list whitelisted container images
+- Skip the garbage collection of the whitelisted container images
+
+### Non-Goals
+
+- Further complicating the kubelet image GC algorithm.
+  - Kubelet already has this logic for the dockershim pause image, we just need
+  to apply it to a configurable list. We do not aim to add any further complexity.
+
+## Proposal
+
+### User Stories
+
+The user stories in the following chapters are alternatives of each other.
+
+#### Story 1
+
+Optimized for small amount of images
+
+I as a Kubernetes administrator would like to define a set of exact image URLs to exclude these images from image garbage collection.
+
+These images may be core system images that would be counter-productive to GC.
+
+#### Story 2
+
+Optimized for big amount of images
+
+I as a Kubernetes administrator have a lots of frequently changing container images and I would like to generally exclude some of them from image garbage collection.
+
+### Implementation Details/Notes/Constraints
+
+What are the caveats to the implementation?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they releate.
+
+#### Details of Story 1
+
+Optimized for small amount of images
+
+It can become difficult to maintain the list of whitelisted images without cluster maintenance in case of changing container images.
+
+This is already implemented in [pr 68549](https://github.com/kubernetes/kubernetes/pull/68549).
+
+#### Details of Story 2
+
+Optimized for big amount of images
+
+There are two alternatives for implementation:
+* Define only the hostname part of the image URL in the command line parameter. So that all images tagged to the same (fake) remote repository will be whitelisted.
+* Define the whitelist with a regular expression. In this case all images matching the regexp will be whitelisted. A function to verify which images are matching would be useful.
+
+### Risks and Mitigations
```

Comment:
this section is lacking.

Reply:
- accidentally writing the wrong image name on the command line may not be noticed until the image is mysteriously gone